drivers: timer: stm32 LPTIM  input clock with a prescaler

FRASTM opened the pull request:
This PR adds a prescaler to the LPTIM input clock to increase the max timeout (0xFFFF lptim counter unit).
With this prescaler, the mcu is able to sleep up to 256 sec w/o wakeup.

With the LSE clock as input (32768Hz),
prescaler = 1 (no division)   one lptim count is  of  0.03 ms and max timeout is 2 seconds
prescaler = 2  one lptim count is  of  0.06 ms and max timeout is 4 seconds
prescaler = 4   -->   one lptim count is  of  0.12 ms and max timeout is 8 seconds
prescaler = 8    -->  one lptim count is  of  0.24 ms and max timeout is 16 seconds
prescaler = 16  -->  one lptim count is  of  0.48 ms and max timeout is 32 seconds
prescaler = 32  -->   one lptim count is  of  0.97 ms and max timeout is 64 seconds
prescaler = 64  -->  one lptim count is  of  1.95 ms and max timeout is 128 seconds
prescaler = 128   one lptim count is  of  3.9 ms and max timeout is 256 seconds


Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: bindings: lptim stm32 has a prescaler entry for the lptim clock

This is a new parameter to divide the LPTIM input clock
by a prescaler, changing the max reachable timeout of the LP timer.
It will divide the LPTIM input clock by 1 (reset value) up to 128.
The lptim configuration register is written with a 3bit value.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
samples: boards stm32 power mgmt nucleo_wb55rg sets a lptim prescaler

configure the nucleo_wb55rg with a prescaler on the LPTIMer input
clock to increase the max sleep duration.
With 32 the input clock is 1024Hz and max reachable timeout is
64 seconds.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
POSITION_VAL  macro is not defined for stm32l0xx so use __CLZ(__RBIT(value)) instead

FRASTM commented:
Will use the LPTIM1 device only and define the clock  freq as a constant of this DTS NODE from the clock source divided by the prescaler

FRASTM committed:
tests: pm: stm32 soc power mgmt for nucleo_l476rg

Give the nucleo_l476rg board and overlay to enable the lptim
so that usecase can run on that target board
This target is a platform candidate for this PM usecase

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
add the configuration so that nucleo_l476rg can run the tests/subsys/pm/power_mgmt_soc/ 
--> PM is one of the tags in the testcase.yaml : this is done through the LPTIM as kernel timer

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -23,18 +22,28 @@
  * - when timeout irq occurs the counter is already reset
  * - the maximum timeout duration is reached with the LPTIM_TIMEBASE value
- * - with prescaler of 1, the max timeout (LPTIM_TIMEBASE) is 2seconds
+ * - with prescaler of 1, the max timeout (LPTIM_TIMEBASE) is 2 seconds
+ *    clock source of 32768Hz (LSE) div by 1 x 0xFFFF
+ * - with prescaler of 128, the max timeout (LPTIM_TIMEBASE) is 256 seconds
+ *    clock source of 32768Hz (LSE) div by 128 x 0xFFFF
```
What about:
```
 * - with prescaler of 1, the max timeout (LPTIM_TIMEBASE) is 2 seconds:
 *    0xFFFF / (LSE freq (32768Hz) / 1)
 * - with prescaler of 128, the max timeout (LPTIM_TIMEBASE) is 256 seconds:
 *    0xFFFF / (LSE freq (32768Hz) / 128)
```

FRASTM replied:
comment changed

FRASTM commented on dts/bindings/timer/st,stm32-lptim.yaml:
```diff
@@ -9,3 +9,20 @@ include:
+      required: false
+      default: 1
+      description: |
+        Clock divider at the input of the lptimer clock, default reset value is not divided
```
description changed

FRASTM commented on samples/boards/stm32/power_mgmt/blinky/boards/nucleo_wb55rg.overlay:
```diff
@@ -0,0 +1,3 @@
+ &lptim1 {
+	st,prescaler = <32>;
```
done

FRASTM commented:
with comment in the samples/boards/stm32/power mgmt/blinky/   nucleo_wb55rg.overlay

mbolivar-nordic commented:
> with comment in the samples/boards/stm32/power mgmt/blinky/ nucleo_wb55rg.overlay

@FRASTM I am not sure what you mean by this, but it sounds like you may want to add some more comments to this file before merging. Is that right?

Also, please rebase.

FRASTM commented:
rebased on c6b05afc92550eea26fb38ac92cfac114f5e5581

erwango commented on dts/bindings/timer/st,stm32-lptim.yaml:
```diff
@@ -10,3 +10,20 @@ include:
+      default: 1
+      description: |
+        Clock divider at the input of the lptimer clock, default reset value is 'not divided'.
+      enum:
```
Please add a comment on the consequences on the tick duration/precision and the fact that `SYS_CLOCK_TICKS_PER_SEC` may require to be tuned for high values (starting from ?).
Also, please provide some example of known working `st,prescaler`/`CONFIG_SYS_CLOCK_TICKS_PER_SEC` combination.
And maybe a indication stating that above X value there is no guarantee to achieve a correct kernel scheduling.

FRASTM replied:
give details in the description

erwango replied:
Let's try a more formal version (please correct if required)
```
        Clock divider at the input of the LPTIM clock, default reset value is 'not divided'.

        Prescaler allows to achieve higher LPTIM timeout and consequently higher core sleep durations,
        but impacts the tick precision. To be used with caution.
        CONFIG_SYS_CLOCK_TICKS_PER_SEC could be used to tune tick duration and gain precision,
        at kernel timing granularity cost.

        For example, when LPTIM is clocked by the LSE (32768Hz) and st,prescaler = <32>:
        LPTIM global timeout is 64 seconds with an increment of 0.97 ms.
        Using CONFIG_SYS_CLOCK_TICKS_PER_SEC = 4096, tick = 0.24 ms, LPTIM precision is 4 ticks.
        Using CONFIG_SYS_CLOCK_TICKS_PER_SEC = 1024, tick = 0.97 ms, LPTIM precision is 1 ticks.
```

FRASTM replied:
ok, description refined

FRASTM committed:
drivers: timer: stm32 adjust lptimer with slow LPTIM clock

Commit to adjust the next_arr with slow LPTIM clock
The next_arr must not be < 0.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -23,18 +22,28 @@
+#define LPTIM_CLOCK_RATIO DT_PROP(DT_NODELABEL(lptim1), st_prescaler)
+#else
+#error "no LPTIM 1 Device"
+#endif
```
#if DT_NODE_HAS_STATUS removed

erwango reviewed (changes requested):
Please fix title for commit that does "Commit to adjust the next_arr with slow LPTIM clock
The next_arr must not be < 0.".
Otherwise LGTM

FRASTM commented:
rebase on 801fe1fa06bd6f8cc2a55079411fd4feac55eff6

andyross commented on dts/bindings/timer/st,stm32-lptim.yaml:
```diff
@@ -10,3 +10,32 @@ include:
         - st,prescaler
         - st,countermode
+
+properties:
```
The prescaler value is computed and set at runtime, it's part of software state, not hardware configuration.  This wants to be a kconfig, doesn't it?

erwango replied:
This prescaler is actually one register of the timer, so to me this is indeed a hardware configuration (we do have the same on similar timer IPs like the ones used for PWM and we also configure them using device tree).
What we're doing here is really an hardware configuration. But indeed, the actual count per tick will then be computed through SYS_CLOCK_TICKS_PER_SEC, which is Kconfig based.
Does it answer your point ?

andyross replied:
You're computing a value at runtime and setting it in the register to change device state based on how the user of the app (and not the owner of the platform code) wants the hardware to behave.  To me that's 100% kconfig's job.

Devicetree is best used to express immutable facts about hardware ("this SOC has an I2C controller at this address",  "this board puts the first SPI clock on pin 41", "this CPU variant has the FPU enabled").  Kconfig is for things that the application wants to tune ("enable this feature", "include this driver", "reserve this many items for this object cache").

There's absolutely some gray area between the two where there's reasonable argument about which to use.  And there are for sure places still in the code where we're using one when we should be using the other (CONFIG_SYS_CLOCK_CYCLES_PER_SEC is actually one of them -- that's a hardware feature, not an app tunable).

But this particular situation seems pretty clear to me.   You support this prescaler so the **application** can make a tradeoff decision between timing precision and idle power draw.  If you as the platform owner can't make the right decision always, it belongs in the app's hands, which means kconfig.

andyross commented:
That's a fair point, though IMHO auto-computing this from the tick rate seems a little clumsy.  Certainly it's not worth a big bikeshed over this.

Most app authors tend to be told things (by me, as often as not) like "The tick rate is just an arbitrary number, you can set it as high as you like as long as your hardware can support the precision and you aren't subject to rollover bugs with long delays".  Routine systems tend to use over-precise values (e.g. 10 kHz is pretty standard, 50 kHz on intel_adsp!) because of this.  If that setting now has power impact, it's going to be surprising to app authors.

My core point is this: kconfig is trivially easy for applications to use, devicetree is very hard for apps.  If apps need to set something, we should strongly prefer kconfig.

erwango commented:
> My core point is this: kconfig is trivially easy for applications to use, devicetree is very hard for apps. If apps need to set something, we should strongly prefer kconfig.

Ok, got it.

About:
> If it that setting now has power impact, it's going to be surprising to app authors.

Unfortunately, it's already what we are doing, but we don't have much choices on STM32 currently. For low power operations, we have two constraints:
A] In Zephyr kernel atm, it's not possible to switch tick source between modes (same tick source required in running and stop modes)
B] counters are available in STOP mode:
  - RTC, allows long sleep duration but it's not meant to be used in runtime (it's actually a time counter, we need to perform conversions all the time)
  - LPTIM, using 32KHz clock: easy to use but it's a 16 bits counter (on 32KHz: Max timeout 2 seconds)

So, today for STM32 users on Zephyr:
- in non PM use cases, Cortex SysTick is used (with default tick rate applied `default 10000 if TICKLESS_KERNEL`)
- in PM use cases, we're using LPTIM as tick source, but tick rate has to be adapted (to 4096) since counter clock is 32KHz

What we're trying to do here is just to extend the LPTIM timeout (because which is legit for PM use cases), playing with its prescaler, but since it has direct impact on tick count, it requires to play with the tick rate (to an extent acceptable by the app, to be fine tuned by the user > we document this in the binding).

Of course, if it were possible to switch tick sources at change of operation mode, we would be able to use the RTC for long sleep periods, be back on Cortex Systick at run time, and we wouldn't need all this.

mbolivar-nordic commented:
> My core point is this: kconfig is trivially easy for applications to use, devicetree is very hard for apps. If apps need to set something, we should strongly prefer kconfig.

Wait, what? No. Hard disagree on this, and frankly I am sick of this sentiment.

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -174,6 +180,8 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
 	else if (next_arr < (lp_time + LPTIM_GUARD_VALUE)) {
 		next_arr = lp_time + LPTIM_GUARD_VALUE;
 	}
+	/* with slow lptim_clock, LPTIM_GUARD_VALUE of 1 is enough */
+	next_arr = next_arr - 1;
```
Houls it be moved to `drivers: timer: stm32 adjust lptimer with slow LPTIM clock` commit?

FRASTM replied:
yes, done

erwango reviewed (changes requested):
One last change and I'm good with it.

FRASTM commented:
rebase and move one change in commit f6445c1a8fa11432695ee353efa7893738bcc813

FRASTM commented:
With high prescaler values (especially above 32) the Tick per sec might be adapted to the lptim clock unit (when running tests/kernel/timer with  CONFIG_PM=y )

FRASTM committed:
drivers: timer: stm32 lptim can divide its input clock freq

This PR will divide the LPTIM clock freq to increase the max timeout.
Only one LPTIM instance is considered for PM timer.
The input freq becomes a fraction of the internal PCLK
source (mainly LSE clock). As the tick per sec does not change,
the minimum lptim counter must always be >0.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

mbolivar-nordic commented on dts/bindings/timer/st,stm32-lptim.yaml:
```diff
@@ -10,3 +10,32 @@ include:
+        Prescaler allows to achieve higher LPTIM timeout (up to 256s when lptim clocked by LSE)
+        and consequently higher core sleep durations, but impacts the tick precision.
+        To be used with caution.
+        CONFIG_SYS_CLOCK_TICKS_PER_SEC could be used to tune tick duration and gain precision,
```
Do you really need this? I know there's not a great place to put driver specific information in our documentation, but this seems like a clear layering violation. DT bindings for an individual SoC timer should be pretty far away from zephyr system clock configuration in the layer, right?

erwango replied:
>  DT bindings for an individual SoC timer should be pretty far away from zephyr system clock configuration in the layer, right?

This is not system clock configuration, but systick rate, which is directly linked because this timer is the actual tick source when PM is used (on the whole stm32 family).

Question is then where to put this information because this prescaler has direct consequences on tick precision (to gain sleep duration), which might need to be counter balanced by a fine tuning of `CONFIG_SYS_CLOCK_TICKS_PER_SEC`.

mbolivar-nordic commented on samples/boards/stm32/power_mgmt/blinky/sample.yaml:
```diff
@@ -10,5 +10,6 @@ tests:
             dt_enabled_alias_with_parent_compat("led0", "gpio-leds") and
-            dt_compat_enabled("st,stm32-lptim")
+            dt_compat_enabled("st,stm32-lptim") and
+            dt_chosen("zephyr,pm-timer")
```
This PR's usage of this chosen node is not acceptable right now.

1. Zephyr chosen nodes are for configuring a piece of hardware used by a subsystem, like the console. This PR seems to introduce a new chosen node that doesn't satisfy this requirement. The commit logs provide no justifications.
2. All such chosen nodes are documented in the table at the bottom of doc/build/dts/api/api.rst. If you can convince me that the requirements for 1. are met, we will need this.

FRASTM replied:
> This PR's usage of this chosen node is not acceptable right now.
> 
> 1. Zephyr chosen nodes are for configuring a piece of hardware used by a subsystem, like the console. This PR seems to introduce a new chosen node that doesn't satisfy this requirement. The commit logs provide no justifications.
> 2. All such chosen nodes are documented in the table at the bottom of doc/build/dts/api/api.rst. If you can convince me that the requirements for 1. are met, we will need this.

1- the chosen node for the `pm-timer` is placed to select which HW peripheral will be used for power-management timer. Some stm32 MCUs have more than one instance of LPTIM. Even if several LPTIM devices are enabled in the DTS, only one device is selected for PM timer : the _chosen_ one.

erwango commented:
To be rebased after merge of https://github.com/zephyrproject-rtos/zephyr/pull/48771

FRASTM commented:
rebase on 3217d9737530bddbd1a109612ec9cdb1765e3ff1 
Choose one lptim instances with `zephyr,pm-timer = &lptim1`

FRASTM commented:
to come after #49357 is merged

FRASTM commented:
rebased on 2e04bfdfe0f8fda3ec5df1365f8f71c744cf5c9f

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -49,66 +72,117 @@ static const struct device *const clk_ctrl = DEVICE_DT_GET(STM32_CLOCK_CONTROL_N
+/* Actual lptim clock freq when the clock source is reduced by the prescaler */
+#define LPTIM_CLOCK (CONFIG_STM32_LPTIM_CLOCK / LPTIM_CLOCK_RATIO)
+
+/* Nb of LPTIM counter unit per kernel tick (depends on lptim clock prescaler) */
+#define COUNT_PER_TICK  (LPTIM_CLOCK / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
```
Not used anymore, please remove

FRASTM replied:
removed

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -49,66 +72,117 @@ static const struct device *const clk_ctrl = DEVICE_DT_GET(STM32_CLOCK_CONTROL_N
+/* The prescaler given by the DTS and to apply to the lptim1 clock */
+#define LPTIM_CLOCK_RATIO DT_PROP(DT_NODELABEL(lptim1), st_prescaler)
+
+/* Actual lptim clock freq when the clock source is reduced by the prescaler */
+#define LPTIM_CLOCK (CONFIG_STM32_LPTIM_CLOCK / LPTIM_CLOCK_RATIO)
```
Should use lptim_clock_freq

FRASTM replied:
removed constant, but _lptim_clock_freq/LPTIM_CLOCK_RATIO_ is used though

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -49,66 +72,117 @@ static const struct device *const clk_ctrl = DEVICE_DT_GET(STM32_CLOCK_CONTROL_N
+ * - with prescaler of 128, the max timeout (LPTIM_TIMEBASE) is 256 seconds:
+ *    0xFFFF / (LSE freq (32768Hz) / 128)
+ */
+#define LPTIM_TIMEBASE CONFIG_STM32_LPTIM_TIMEBASE
```
Not used anymore

FRASTM replied:
removed

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -204,9 +278,9 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
-	/* add count unit from the expected nb of Ticks */
-	next_arr = next_arr + ((uint32_t)(ticks) * lptim_clock_freq)
-			/ CONFIG_SYS_CLOCK_TICKS_PER_SEC - 1;
+	next_arr = next_arr + ((uint32_t)(ticks) * LPTIM_CLOCK)
```
```suggestion
	next_arr = next_arr + ((uint32_t)(ticks) * lptim_clock_freq)
```

FRASTM replied:
_lptim_clock_freq/LPTIM_CLOCK_RATIO_ indeed in that formula

FRASTM commented:
rebase

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -67,6 +67,9 @@ static int32_t lptim_time_base;
 #define LPTIM_CLOCK_RATIO DT_PROP(DT_NODELABEL(lptim1), st_prescaler)
 
+/* Actual lptim clock freq when the clock source is reduced by the prescaler */
+static uint32_t lptim_clock = 32000;
```
Use of `lptim_clock` name vs `lptim_clock_freq` is confusing.
This does not explain the difference between the 2 values, and why sometime `lptim_clock_freq / LPTIM_CLOCK_RATIO` should be used instead of `lptim_clock_freq`.

What about `lptim_set_autoreload((lptim_clock_freq / CONFIG_SYS_CLOCK_TICKS_PER_SEC) - 1);`

erwango commented on samples/boards/stm32/power_mgmt/blinky/sample.yaml:
```diff
@@ -10,5 +10,6 @@ tests:
             dt_enabled_alias_with_parent_compat("led0", "gpio-leds") and
-            dt_compat_enabled("st,stm32-lptim")
+            dt_compat_enabled("st,stm32-lptim") and
+            dt_chosen("zephyr,pm-timer")
```
References to `zephyr,pm-timer` to be removed as it's not used anymore

FRASTM replied:
removing the commits  related to the the multi-instance of lptim and the method for choosing one.

erwango reviewed (changes requested):
Please drop the multi instance support for this PR.

FRASTM commented:
Rebase on c4b10874eb9bd46b9b824d16aa25545dc2f788e6
Reduce the lptim_clock_freq by LPTIM_CLOCK_RATIO during the init.

FRASTM committed:
drivers: timer: stm32u5 lptimer waits for DIER  complete

On the stm32U5, when modifying the DIER register of the LPTIM peripheral,
a new write operation to can only be performed when the previous write
operation is completed and before going-on.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented:
@FRASTM please rebase

FRASTM commented:
rebase on 6c1363e34dfc121fed3346c55241fcfd87b2070a

mbolivar-nordic commented:
Not happy about the Kconfig option and I think it could be moved elsewhere, but accepted for now since I don't have a concrete suggestion.

@kartben this is an example of where it would be useful to have an enumeration of device drivers and associated docs.
